### PR TITLE
[electrophysiology_browser] Replace Physiological Annotations with Events

### PR DESCRIPTION
**Pull request by jeffersoncasimir**
This PR aims to improve the current compliance to the BIDS standard. Loris-MRI counterpart: [#1041](https://github.com/aces/Loris-MRI/pull/1041)
Partially addresses #8910.

This replaces references to `annotation` files with references to `event` files.

This PR deletes `*_annotation_*` tables. The corresponding required event tables already exist.

+minor improvements to the `EventManager` and `AnnotationForm` components.

This PR brings in the following limitations:
- Events can no longer be created via the interface (support coming soon)
- Events can be modified, but the currently supported fields are by default `read-only` or `disabled` (Name and time)
- HED tag support is extremely limited (better support in imminent PR: #9033)

~*TODO: Raisinbread modifications*~

TODO: Get official server-side solution + Add to documentation.
- [ ] Determine required request time limit
- [ ] Determine required memory allocation limit
- [ ] Determine procedure for increasing when needed

Both of the above needed to be increased for certain requests.

### CODE REVIEW

**Comment by driusan** [comment]
@regisoc can you review this?

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc [hunk 1]
```diff
@@ -57,7 +78,340 @@ class ElectrophysioEvents
...
+        $db = \NDB_Factory::singleton()->database();
+
+        $physioFileID = $db->pselectone(
+            'SELECT PhysiologicalFileID
+              FROM physiological_task_event
+              WHERE PhysiologicalTaskEventID=:taskEventID',
+            ['taskEventID' => $physiologicalTaskEventID]
+        );
+
+        // TODO: Check that this cascades properly to rel tables
```
Comment (regisoc) [inline modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:184]: Does it? Else, it is missing some delete.

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc [hunk 2]
```diff
@@ -57,7 +78,340 @@ class ElectrophysioEvents
...
+        //Get data directory base path from Config
+        $dataDir = $db->pselectOne(
+            'SELECT Value
+            FROM Config AS config
+            INNER JOIN ConfigSettings AS c
+            ON c.Name=:name AND config.ConfigID=c.ID',
+            ['name' => 'dataDirBasepath']
+        );
+
+        $tsv = $db->pselect(
```
Comment (regisoc) [inline modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:213]: As a note, not sure if it is the intended behaviour, use `pselectRow` instead to remove the need for `$tsv[0]` after?

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc [hunk 3]
```diff
@@ -57,7 +78,340 @@ class ElectrophysioEvents
...
+      FROM physiological_event_file
+      WHERE PhysiologicalFileID=:PFID
+      AND FileType='tsv'",
+            ['PFID' => $this->_physioFileID]
+        );
+
+        $tsvPath = count($tsv) > 0 ? $dataDir . $tsv[0]['filePath'] : '';
+
+        // Update files if files updated before database updated
+        if ($tsv[0]['lastWritten'] <= $tsv[0]['lastUpdate']) {
```
Comment (regisoc) [inline modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:229]: We should check the validity of `$this->_physioFileID`. Maybe in constructor? 
If `$this->_physioFileID` does not exist in db, `$tsv` is empty, and I think this should trigger a warning here (`$tsv[0]` does not exist).

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc [hunk 4]
```diff
@@ -20,34 +21,54 @@ class ElectrophysioEvents
...
+                ON f.EventFileID = te.EventFileID
                 WHERE f.PhysiologicalFileID=:PFID AND f.FileType="tsv"',
             ['PFID' => $this->_physioFileID]
         );
 
-        /**
-         * TODO: Get event params and metadata.
-         *       NOT in the scope of current task
-         **/
+        $taskEventIDs = array_map(
```
Comment (regisoc) [inline modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:40]: Open question: do we need to check first that `$taskEvents` is not empty, or that the given `$physioFileID` exists in db? Let say we try to build an `electrophysioevents` with a `$physioFileID` that is set to a value that does not exist in db, is that an expected behaviour?

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc [hunk 5]
```diff
@@ -57,7 +78,340 @@ class ElectrophysioEvents
...
+            );
+
+            // TODO: Make columns more dynamic
+            $tsvEntries = [
+                'onset', 'duration', 'sample', 'trial_type', 'response_time', 'value'
+            ];
+            foreach ($columnNames as $columnName) {
+                $tsvEntries[] = $columnName['PropertyName'];
+            }
+            //            $tsvEntries[] = 'HED';
```
Comment (regisoc) [inline modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:262]: I guess we should let it there for future HED integration.

File: modules/electrophysiology_browser/php/events.class.inc [hunk 6]
```diff
@@ -89,53 +99,53 @@ class Annotations extends \NDB_Page
...
 
             return (new \LORIS\Http\Response\JSON\OK());
         case 'POST':
-            $parameters = json_decode((string) $request->getBody(), true);
-            if (!$user->hasPermission('electrophysiology_browser_edit_annotations')
+            // TODO: Better failure reporting
+            $parameters = json_decode((string)$request->getBody(), true);
+
+            if (!$user->hasPermission(
+                'electrophysiology_browser_edit_annotations'
```
Comment (regisoc) [inline modules/electrophysiology_browser/php/events.class.inc:111]: Rename permission?

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx [hunk 7]
```diff
@@ -181,17 +162,16 @@ const AnnotationForm = ({
    */
   const handleReset = () => {
     // Clear all fields
-    setEvent(['', '']);
-    setTimeSelection([null, null]);
-    setLabel('');
-    setComment('');
+    // setEvent(['', '']);
+    // setTimeSelection([null, null]);
+    // setLabel('');
```
Comment (regisoc) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx:167]: Remove comments?

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx [hunk 8]
```diff
@@ -181,17 +162,16 @@ const AnnotationForm = ({
...
   };
 
   /**
    *
    */
   const handleDelete = () => {
-    setIsDeleted(true);
+    // setIsDeleted(true);
```
Comment (regisoc) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx:174]: Remove comment?

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx [hunk 9]
```diff
@@ -432,57 +433,70 @@ const AnnotationForm = ({
...
-          }
+          {/*{currentAnnotation &&*/}
+          {/*  <button*/}
+          {/*    type="button"*/}
+          {/*    onClick={handleDelete}*/}
+          {/*    className="btn btn-primary btn-xs"*/}
+          {/*  >*/}
+          {/*    Delete*/}
+          {/*  </button>*/}
+          {/*}*/}
```
Comment (regisoc) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx:499]: remove comments?

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.tsx [hunk 10]
```diff
@@ -1258,58 +1237,33 @@ const SeriesRenderer: FunctionComponent<CProps> = ({
...
+                {/*        : setRightPanel('annotationForm');*/}
+                {/*      setCurrentAnnotation(null);*/}
+                {/*    }}*/}
+                {/*  >*/}
+                {/*    {rightPanel === 'annotationForm'*/}
+                {/*      ? 'Close Annotation Form'*/}
+                {/*      : 'Add Annotation'*/}
+                {/*    }*/}
+                {/*  </button>*/}
+                {/*}*/}
```
Comment (regisoc) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.tsx:1257]: Remove comments?

**Review by regisoc — COMMENTED** [review]
Nice work! 
- global logic looks good. Some questions for the back-end class.
- Still some TODOs in the code. I avoided repeating all comments. The comment on the first one should be valid for all similar following it.
- rename permission `electrophysiology_browser_edit_annotations` to `electrophysiology_browser_edit_events`?
- Some cleaning to do (commented lines to remove/keep)?
- Not tested yet. I might need help setting up testing env.

File: modules/electrophysiology_browser/php/events.class.inc [hunk 11]
```diff
@@ -32,6 +34,10 @@ class Annotations extends \NDB_Page
 
         switch ($request->getMethod()) {
         case 'GET':
+            // TODO: Get official server-side solution + Add to documentation
```
[the remark this answers is not in the document]
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/events.class.inc:37]: I added this to the PR description

File: modules/electrophysiology_browser/php/events.class.inc [hunk 12]
```diff
@@ -32,6 +34,10 @@ class Annotations extends \NDB_Page
...
+            // set_time_limit(300);  // Increase request time limit to 5 minutes
+            // ini_set('memory_limit', '1G'); // Increase memory allocation limit
```
[the remark this answers is not in the document]
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/events.class.inc:39]: I needed to increase these values on my VM in order to use certain features.

Resolving by adding to the PR description as tech debt.

File: modules/electrophysiology_browser/php/events.class.inc [hunk 13]
```diff
@@ -73,12 +79,16 @@ class Annotations extends \NDB_Page
...
             return $downloader->handle(
                 $request->withAttribute('filename', $filename)
             );
         case 'DELETE':
-            $parameters = json_decode((string) $request->getBody(), true);
-            if (!$user->hasPermission('electrophysiology_browser_edit_annotations')
+            $parameters = json_decode((string)$request->getBody(), true);
+
+            if (!$user->hasPermission(
+                'electrophysiology_browser_edit_annotations'
```
[the remark this answers is not in the document]
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/events.class.inc:90]: The decision was made to not change the permission name at this time.

More significant changes to the permission (ex: granularity) are required for this module in the near future.

File: modules/electrophysiology_browser/php/events.class.inc (hunk 6 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/events.class.inc:111]: Will not be done at this time (see first mention)

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc [hunk 14]
```diff
@@ -57,7 +78,340 @@ class ElectrophysioEvents
...
+        array $instance_data,
+        array $metadata,
+        ?int $instance_id,
+    ): array {
+
+        $factory = \NDB_Factory::singleton();
+        $user    = $factory->user();
+        $db      = $factory->database();
+
+        if ($user->hasPermission('electrophysiology_browser_edit_annotations')) {
```
[the remark this answers is not in the document]
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:101]: Will not be done at this time (see first mention)

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc [hunk 15]
```diff
@@ -57,7 +78,340 @@ class ElectrophysioEvents
...
+                ['PFID' => $this->_physioFileID]
+            );
+
+            //Get data from POST request
+            $metadata = [
+                'Description' => $metadata['description'],
+                'Sources'     => $metadata['sources'],
+                'Author'      => $metadata['author']
+            ];
+            error_log(json_encode($metadata));
```
[the remark this answers is not in the document]
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/electrophysioevents.class.inc]: Removed (no update support)

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc (hunk 2 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:205]: This was copied verbatim from `electrophysioannotations.class.inc`

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx (hunk 8 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx]: Removed line. Unsupported

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx (hunk 7 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx]: Removed; unsupported

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx (hunk 9 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/AnnotationForm.tsx:496]: Removed; unsupported

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.tsx (hunk 10 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.tsx]: Removed; unsupported

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc (hunk 1 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:176]: Comment removed. This is related to the `physiological_task_event_opt` and `physiological_task_event_hed_rel` tables with `physiological_task_event` FKs

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc (hunk 3 above)
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/electrophysioevents.class.inc]: Good find. The logic now addresses this corner case.

**Comment by regisoc** [comment]
I will put that here so we do not forget them. Note after tests:
- Event download button disabled or not working.
- Error `epoch.properties is undefined` following that process:
  - open event panel
  - edit one event
  - submit this event
  - close the event submit with the top-right "x".

**Comment by regisoc** [comment]
@jeffersoncasimir let me know if you need another round of test this morning.

**Review by regisoc — CHANGES_REQUESTED** [review]
New pass: 
- error on `epoch` resolved.
- new errors due to multiple config db calls that can vary from config.xml. Unify calls for config check to `$config->getSetting("dataDirBasepath");` in electrophysiologyevents.class.inc (one in `_updateArchives` one in `update`).
- ~~Archive events file `tgz` created and downloadable, but archive is corrupted.~~ Not corrupted, but not a gzip archive? Using `tar xf <file>` works, but not when using `tar xzf <file>`?

**Review by regisoc — APPROVED** [review]
LGTM
